doozerlib/cli/release_gen_payload.py

```diff
@@ -170,16 +189,32 @@ def _get_payload_images(self, images):
 
         return payload_images, invalid_name_items
 
-    def _get_latest_builds(self, payload_images):
+    def _get_latest_builds(self, payload_images: List[ImageMetadata]) -> Tuple[List[BuildRecord], List[ImageMetadata]]:
         """
-        find latest brew build (at event, if given) of each payload image.
+        Find latest brew build (at event, if given) of each payload image.
+
+        If assemblies are disabled, it will return the latest tagged brew build with the candidate brew tag.
+        If assemblies are enabled, it will return the latest tagged brew build in the given assembly. If no such build, it will fall back to the latest build in "stream" assembly.
+
         :param payload_images: a list of image metadata for payload images
         :return: list of build records, list of images missing builds
         """
         tag_component_tuples = [(image.candidate_brew_tag(), image.get_component_name()) for image in payload_images]
-        brew_latest_builds = brew.get_latest_builds(tag_component_tuples, "image", self.brew_event, self.brew_session)
-        # there's zero or one "latest" build in each list; flatten the data structure.
-        brew_latest_builds = [builds[0] if builds else {} for builds in brew_latest_builds]
+        lists_of_brew_builds = brew.get_tagged_builds(tag_component_tuples, "image", self.brew_event, self.brew_session)
```

Comment:
I found this API that allows specifying globs to match NVRs which may help cut down on the data transfer:
https://koji.fedoraproject.org/koji/api

```
listBuilds(packageID=None, userID=None, taskID=None, prefix=None, state=None, volumeID=None, source=None, createdBefore=None, createdAfter=None, completeBefore=None, completeAfter=None, type=None, typeInfo=None, queryOpts=None, pattern=None)
  Return a list of builds that match the given parameters

        Filter parameters
            - packageID: only builds of the specified package (numeric id)
            - userID: only builds owned by the given user (numeric id)
            - taskID: only builds with the given task ID
                If taskID is -1, only builds with a non-null task id
            - volumeID: only builds stored on the given volume (numeric id)
            - source: only builds where the source field matches (glob pattern)
            - prefix: only builds whose package name starts with that prefix
            - pattern: only builds whose nvr matches the glob pattern
            - state: only builds in the given state (numeric value)
```

Using a glob in `pattern` we might more efficiently query. 

1. Search for `*.assembly.{name}` 
2. If no results, search for, `*.assembly.stream`  

Getting all tagged builds for all images could return tens-of-thousands of records for an aging release. Thoughts?

I enabled brew-event pinning for the API here: https://github.com/openshift/doozer/pull/413.

Reply:
I am trying the `listBuilds` and also comparing the numbers. Will put my findings later.